### PR TITLE
[CFX-7556] Scope per-framework PR triggers to relevant changed files

### PR DESCRIPTION
**Pull request by taras-pokornyy**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary

- Scopes the 8 `*_local_on_pr` Harness triggers (java_codegen, python3_keras/onnx/pytorch/sklearn/xgboost, python312, r_lang) so they only fire when the diff touches `custom_model_runner/` (DRUM source), the framework's own env dir, `tests/`, `harness_scripts/functional_by_framework/`, or `requirements_for_per_framework_tests.txt` — previously they had no path filter and ran on every PR regardless of what changed.
- Fixes the CI bloat described in [CFX-7556](https://datarobot.atlassian.net/browse/CFX-7556): PRs touching only `public_dropin_notebook_environments/python313_notebook/**` (e.g. [#2304](https://github.com/datarobot/datarobot-user-models/pull/2304)) were triggering the full sklearn/R/Java functional test matrix, including flaky suites unrelated to the change.
- DRUM-source and shared-fixture coverage is preserved: since `test_functional_by_framework` installs `datarobot-drum` from `custom_model_runner/` into every framework's image, the condition is an OR across DRUM source + that framework's path + shared test infra, not just the framework's own directory — a `custom_model_runner/**` change still fires all 8.

## Rationale

[CFX-7556]: https://datarobot.atlassian.net/browse/CFX-7556?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Low Risk**
> CI trigger-only change; wrong path patterns could skip needed runs, but the shared-path OR matches existing push-trigger patterns in `.harness`.
> 
> **Overview**
> Adds **`changedFiles` regex filters** to the eight `*_local_on_pr` Harness GitHub PR triggers so each framework’s `test_functional_by_framework` run only fires when the PR touches paths that matter for that stack—not on every PR to `master`.
> 
> Each trigger now requires (in addition to `targetBranch: master`) at least one changed file matching an OR of **`custom_model_runner/`**, that trigger’s **`public_dropin_environments/<framework>/`** tree, shared **`tests/`**, **`harness_scripts/functional_by_framework/`**, **`requirements_for_per_framework_tests.txt`**, and for **python312** also **`model_templates/`**. Unrelated areas (e.g. notebook-only env dirs) should no longer spin up the full per-framework matrix.
> 
> **`custom_model_runner/`** remains in every regex so DRUM source changes still run all eight suites, matching how functional tests install DRUM from that tree.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 798e97c89dd0b7f064d192d32ee22101840857ea. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .harness/java_codegen_local_on_pr.yaml [hunk 1]
```diff
@@ -22,6 +22,9 @@ trigger:
             - key: targetBranch
               operator: Equals
               value: master
+            - key: changedFiles
+              operator: Regex
+              value: ^(custom_model_runner/|public_dropin_environments/java_codegen/|tests/|harness_scripts/functional_by_framework/|requirements_for_per_framework_tests\.txt)
```
[the remark this answers is not in the document]
Comment (taras-pokornyy) [reply .harness/java_codegen_local_on_pr.yaml]: Fixed in commit : [1dd1981](https://github.com/datarobot/datarobot-user-models/pull/2371/commits/1dd19811f58e5c15020bd0333cffc2df5784a456)

File: .harness/python3_sklearn_local_on_pr.yaml [hunk 2]
```diff
@@ -22,6 +22,9 @@ trigger:
             - key: targetBranch
               operator: Equals
               value: master
+            - key: changedFiles
+              operator: Regex
+              value: ^(custom_model_runner/|public_dropin_environments/python3_sklearn/|tests/|harness_scripts/functional_by_framework/|requirements_for_per_framework_tests\.txt).*
```
[the remark this answers is not in the document]
Comment (taras-pokornyy) [reply .harness/python3_sklearn_local_on_pr.yaml:27]: Verified against the actual test code before fixing:

- `task_templates/` isn't used anywhere in executable test code — the only reference in `tests/functional/test_inference_per_framework.py` is inside a comment (lines ~441-449) explaining how `drum fit` works, not a live fixture load. No fix needed for that path.
- `model_templates/` IS genuinely used, but only by `TestPython312Fips.test_predict` in `tests/functional/test_inference_per_framework.py` (`os.path.join(MODEL_TEMPLATES_PATH, "python3_dummy_regression")`), which is gated by `skip_if_framework_not_in_env(PYTHON312, framework_env)` — it only actually runs for python312.

So I've added `model_templates/` to `python312_local_on_pr.yaml`'s regex (b5cf6998). I'm leaving `python3_sklearn_local_on_pr.yaml` and `r_lang_local_on_pr.yaml` unchanged — there's no code path where their actual test run touches `model_templates/` or `task_templates/`, so adding those paths there would just be unverified scope.

**Review by peterzdeb — APPROVED** [review]

**Review by peterzdeb — COMMENTED** [review]
Reviewed the updated version. Dropping the seven `env_info.json` files resolved the conflict with #2374 — that was the one I'd have blocked on, since resolving it the wrong way would have left `tags` pointing at superseded environment version ids. Now `+24/-0` across the 8 trigger files, and the core design looks right: `custom_model_runner/` leads every pattern, so a DRUM-source change still fans out to all eight frameworks, which is the correctness-critical part.

Two things I'd like to see settled before this lands. Neither is a blocker.

### 1. `model_templates/` is in python312's pattern but not the other seven

```
python312     …|harness_scripts/functional_by_framework/|model_templates/|requirements…
java_codegen  …|harness_scripts/functional_by_framework/|requirements…          ← no model_templates/
keras · onnx · pytorch · sklearn · xgboost · r_lang                              ← same
```

Each of those frameworks has its own fixture directory:

```
model_templates/python3_keras          model_templates/python3_sklearn
model_templates/python3_onnx_*         model_templates/python3_xgboost
model_templates/python3_pytorch        model_templates/java_codegen
model_templates/r_lang
```

`tests/` covers changes to test *code*, but not to these fixtures. So as written, a PR editing only `model_templates/python3_sklearn/**` would not run the sklearn functional suite.

This looks like it has to be wrong in one direction or the other:

- if python312's inclusion is intentional, the other seven need `model_templates/` too, otherwise they lose fixture coverage
- if it isn't, python312 should drop it — right now any `model_templates/` change fires python312, including `model_templates/r_lang`

Happy either way, but the asymmetry seems unintended.

### 2. First use of `operator: Regex` on `changedFiles` in this repo — worth confirming the semantics

`grep -rl "operator: Regex" .harness/` currently returns nothing, so this is the first one, and the any-vs-all behaviour is unverified here.

Worth checking because `changedFiles` semantics do vary by operator in ways the YAML doesn't hint at. In `global-envs-models` there are two triggers on the same repo watching the same file, one with `Equals` and one with `In` — the `In` one fires on a multi-file commit, and the `Equals` one does not. That difference has meant one of its two images silently never rebuilt on release branches.

If `Regex` turns out to require **every** changed file to match rather than **any**, the effect here isn't a narrower trigger — it's no trigger at all for any mixed PR (framework file plus a README), which would be a silent loss of gating rather than the intended reduction.

Cheap to de-risk: land one trigger, push a throwaway two-file PR (one matching path + one non-matching), confirm it fires, then land the other seven.

### Minor, non-blocking

- Nothing in the patterns covers the trigger YAML itself, so editing `.harness/<framework>_local_on_pr.yaml` won't exercise that framework's tests. Probably fine, just noting CI config changes aren't self-testing.
- `tests/` is broad enough that any test edit still fans out to all eight, which trims less than the description implies — though that's the safe direction to err.

One thing worth knowing given the timing: these eight pipelines gate the execution-environment work, and a CVE-driven environment bump touches seven `env_info.json` files in a single commit. If the semantics in point 2 turn out to be "all files must match", those bumps would stop triggering these suites — another reason to confirm it on one trigger first.

**Comment by taras-pokornyy** [comment]
@peterzdeb, thanks for the thorough review — both points are fair to press on given the stakes here.

### 1. `model_templates/` asymmetry

Checked this directly rather than guessing. It's intentional, not a bug: **only python312 actually touches `model_templates/`** in this test run. `TestPython312Fips.test_predict` (`tests/functional/test_inference_per_framework.py:998`) does `os.path.join(MODEL_TEMPLATES_PATH, "python3_dummy_regression")` and is gated on `PYTHON312` only.

I searched all 5 shared test files (`test_inference_per_framework.py`, `test_inference_gpu_predictors.py`, `test_fit_per_framework.py`, `test_other_cases_per_framework.py`, `test_unstructured_mode_per_framework.py`) plus r_lang's 3 extra files (`test_language_predictors.py::TestRPredictor`, `test_drum_utils.py`, `test_model_metadata.py`) and `tests/conftest.py`. The other 7 frameworks' own `model_templates/<framework>` directories exist, but nothing in this suite references them — there's no generic per-framework helper like `os.path.join(MODEL_TEMPLATES_PATH, framework_env)`. The only other `MODEL_TEMPLATES_PATH` usages (in `test_inference_gpu_predictors.py`, for `triton_onnx_unstructured`/`gpu_nim_textgen`/`gpu_vllm_textgen`) are gated on GPU-only framework constants that never match any of the 8 frameworks this job runs — so those are unconditionally skipped here too.

So the asymmetry is correct as written. No change needed.

### 2. Regex any-vs-all semantics — agreed, needs empirical confirmation

You're right not to take this on faith, especially given the `global-envs-models` precedent. I got an answer from an AI assistant earlier claiming any()-existential semantics for `changedFiles` operators generally (including Regex), which is what the trailing-`.*` fix assumed — but that was never verified against our actual Harness instance, and I don't want to treat an AI-agent's assertion as ground truth for something with this blast radius.

Plan: once this PR is merged (trigger `payloadConditions` are read from the branch they're pinned to — `master` — so this can't be tested from the PR branch itself), I'll open a throwaway draft PR touching exactly two files — one matching a single trigger's scope (e.g. `public_dropin_environments/python3_sklearn/README.md`) and one matching nothing (root `README.md`) — and confirm via commit statuses whether the sklearn trigger fires. That directly answers the any-vs-all question and specifically covers the multi-`env_info.json`-file CVE-bump scenario you flagged.

**Review by peterzdeb — APPROVED** [review]
